### ISSUE TITLE
Image rendering is broken

### ISSUE THREAD

**Issue by vvavepacket** [brh55/react-native-masonry#70]
Hi,

I tried 0.4.4 release, but noticed that the rendering is still broken. I am loading my images via the internet (not local images). Could this be the reason?

![image](https://user-images.githubusercontent.com/8295642/34197764-dee293e2-e535-11e7-8a16-66088e1b2c4d.png)

**Comment by brh55** [comment]
Can you let me know which version of React + RN are you on? I'm not sure if it's an update with RN or not

**Comment by vvavepacket** [comment]
Here are my react versions

    "react": "16.0.0-alpha.12",
    "react-native": "0.45.1",

**Comment by jozekcore** [comment]
i have the same problem, help please

**Comment by vvavepacket** [comment]
@brh55  let us know if you need any information or anything we can provide you. thank you

**Comment by brh55** [comment]
If someone can replicate a basic repo, it would be easier to debug the root issue of this problem.

**Comment by mckrava** [comment]
Have the same issue with next settings:
`"react": "^16.1.1"`
 `"react-native": "v0.49.3 "`

**Comment by mckrava** [comment]
As a fast dirty solution I refresh the component after images have been loaded. I do it by settings columns to `this.state.column = 1` in `constructor()` and calling setState({column: 2}) after a timeout in `ComponentDidMount`.  But it's a madness. Please, fix.

**Comment by vvavepacket** [comment]
@brh55 Where can I send my private repo? Thank you.

**Comment by vvavepacket** [comment]
@brh55 

Below is youtube video. I have applied mckrava trick. Here, we can notice that when the react native component is re rendered again (clicking back to the previous screen), the images are messed up.

[YouTube bug video](https://www.youtube.com/watch?v=CqLQGbqZwpU)

Below is zipped source code:

[Source code](https://drive.google.com/open?id=1fQg2s4RJQDCbRXKVH9tKiA7EBUzTQ2jn)

Please kindly take a look and let us know how can we help resolve issue. Thank you.

**Comment by mckrava** [comment]
@vvavepacket Strangely, in my case all images render properly after any number of reloads.

Here is my [code](https://drive.google.com/file/d/1IuvE3ojg1t37duEBIrlS2Jwe2c4q9P1A/view?usp=sharing)

**Comment by vvavepacket** [comment]
@mckrava  I see

@brh55 

I found some obervation. it seems the image is broken if masonry detectes the internet image has been cache.. Now, when i try to append the end of the image URI with a random number such as this

`var img =  'https://i.imgur.com/zYwYIt1.png' + '?' + new Date().getTime(),`

The rendering is fine,, except that due to the image URI being different, it fetches it always thus it is slow. I believe the issue is with caching.

**Comment by brh55** [comment]
@vvavepacket 
I'm not sure if you are using the repo provided for your current source, but I've noticed multiple instances where you are changing state between different component stages, which is probably causing the issues with RN-Masonry.

In your current components:
```js
    componentWillMount() {
        this.props.fetchRegionList();

        this.setState({
            column: 2,
            borderRadius: 5
        })
    }

    componentDidMount() {
        this.setState({
            column: 3,
            borderRadius: 5
        })
    }

    componentWillUnmount() {
        this.setState({
            column: 2,
            borderRadius: 5
        })
    }
```
In this case you'll want to set the state within the class `constructor` like such:

```js
class D_RegionList extends Component {
	constructor() {
                super();
		this.state = {
			column: 3,
			borderRadius: 10,
		};
	}

       // Place any fetching/side-effects within  componentDidMount()
       componentDidMount() {
		this.props.fetchRegionList();
      }

      // You can remove unmount code most likely
      // omitted further code for condenseness
```
Hence, why we are seeing some weird behaviors when combined with using seeing some issues with it loading 2 columns when going back to a previous scene. Now with those revisions, Masonry should behave as expected.

....
Needless to say `Masonry` should have behaved properly even with the constant state changes. And I've realized the issue with the existing design, which I'll fix shortly.

To explain, in this particular instance masonry is prepping the view for 2 columns (within "willMount") and resolving images, but the component state is then updated ("didMount") for 3 columns, so masonry does not resize any images in between the state changes. This causes different re-size images (some for 3 columns, and other for 2 columns). It probably should have just resize all of the images.

**Comment by vvavepacket** [comment]
@brh55 

Thanks for the explanation, will try your changes.

Is there any way to remove/reset the caching used by Masonry (when going back/forth/reload screens)?

**Comment by vvavepacket** [comment]
@brh55 

Fantastic. I tried your solution and it resolves the 2 or 3 columns issue. Regarding the caching, I noticed that if an internet image is already fetched by Masonry, when attempting to render it again by either backing or entering the screen, the layout is broken. 

But when we use a different image URL usually by appending a random number after the URL (e.g. http://yahoo.com/logo.jpg?4374378347), then it appears Masonry appears to redownload the image and get the dimensions again and it renders properly. Do you know why this occurs?


If we remove the URL randomizer string, on first load it looks fine
![image](https://user-images.githubusercontent.com/8295642/34464106-53600aa2-ee40-11e7-957e-92a6c07638f2.png)

Now if we press back, and go to Masonry again, the result looks like this because the image is already downloaded in cache. 

![image](https://user-images.githubusercontent.com/8295642/34464111-7a4289ba-ee40-11e7-9977-5d9ce009de75.png)



I think Masonry is not calculating the dimensions properly or is unable to resolve images in cache?

Thank you.

**Comment by CarlosArango** [comment]
im having the same problem 
react-native-cli: 2.0.1
react-native: 0.51.0

it has ocurred the same, the first time render all images, but  when change the view and comeback the images don't render

**Comment by vvavepacket** [comment]
Are you using internet images? What happens if you will append some random string after image link so that it will not be cached?


Sent via the Samsung GALAXY S®4, an AT&T 4G LTE smartphone

<div>-------- Original message --------</div><div>From: CarlosArango <notifications@github.com> </div><div>Date:01/02/2018  1:19 PM  (GMT-05:00) </div><div>To: brh55/react-native-masonry <react-native-masonry@noreply.github.com> </div><div>Cc: vvavepacket <ivanmatala@gmail.com>, Mention <mention@noreply.github.com> </div><div>Subject: Re: [brh55/react-native-masonry] Image rendering is broken (#70) </div><div>
</div>im having the same problem
react-native-cli: 2.0.1
react-native: 0.51.0

it has ocurred the same, the first time render all images, but when change the view and comeback the images don't render

**Comment by CarlosArango** [comment]
@vvavepacket  
when i add a random string after image link, it is duplicated
<img width="578" alt="captura de pantalla 2018-01-02 a la s 5 40 19 p m" src="https://user-images.githubusercontent.com/4949126/34502962-136e1b1e-efe4-11e7-9527-c0d5e5763c9c.png">

**Comment by brh55** [comment]
@vvavepacket Could you remove LOC 62 - 77 from Masonry.js within your `node_modules/react-native-masonry` and see if this fixes your issue. I'm nearly certain I just need to re-write the caching portion. It uses URI as a means of checking if it needs to re-download + resize images.

**Comment by brh55** [comment]
Closing, added new caching logic in place that should be optimized for existing data, and fix any issues with view changes. Please note, if you are experiencing flickering between columns counts, please check implementation of parent component to ensure masonry isn't being loaded with different initial columns, and switching to other columns.

**Comment by vvavepacket** [comment]
@brh55 

I just pulled 0.4.5 and can say the caching issue is still there (I removed my URI image randomizer). I assume removing LOC 62-77 is the same as 0.4.5?

**Comment by vvavepacket** [comment]
How do we reopen this issue? Thank you.

**Comment by brh55** [comment]
@vvavepacket There shouldn't be any issues at this point, most of the caching logic has been completely removed *(relying on default caching mechanism provided by Image)* with the only exception of processing the difference of blocks when you have the same column count.

I would suspect some issues with the parent component implementation. Could you provide your latest code?

**Comment by vvavepacket** [comment]
@brh55 

[YouTube video detailing caching issue in latest 0.4.5](https://youtu.be/0MyOMo0bEz4)

[Source Code](https://drive.google.com/file/d/1rBXox-mcxBXm3I-kRTaYIKVZeyi-lvab/view?usp=sharing)

**Comment by brh55** [comment]
@vvavepacket I'm thinking since it's nested within a navigator, it may have to do with something ListView not being rerender upon entering because it's already recognized as being loaded with that exact same dataset. Can you try changing in `Masonry` line 20.

`const containMatchingUris = (r1, r2) => isEqual(r1.map(brick => brick.uri), r2.map(brick => brick.uri));`

to 

`const containMatchingUris = () => false;`

**Comment by vvavepacket** [comment]
Here is the result. I think we are close to solving this.

https://youtu.be/pDIXj2Gxv3g

It helps, but we can still see when the user goes back the rendering is
broken initially yet the screen fixes itself immediately. How can we fix
that part?

**Comment by vvavepacket** [comment]
@brh55 

I tried changeing the above containMatchingUris and below is the result. We can see that the image is broken at first, and then it suddenly fixes itself. Do you have any idea how to fix this? **remove the broken rendering at first glance?**

[First render bug](https://www.youtube.com/watch?v=HAsbCAIPfQQ)

**Comment by brh55** [comment]
@vvavepacket

I think the issue stems from the initial state rendering without any parent dimensions, but it fixes itself when the `state.dimensions.width` updated. 

Can you can set the masonry's component `dimension.width` properties with your screens width, and see if that fixes the flashing of the rendering. If this is the case, we can come up with some ways to cover up the issue.

`Masonry.js`
LOC: 47
```js
this.state = {
    dataSource: this.ds.cloneWithRows([]),
   dimensions: {
      width:  // Your device masonry width here
   },
    initialOrientation: true,
    _sortedData: [],
    _resolvedData: []
};
```

**Comment by hoscarcito** [comment]
I'm facing the same, but removing the function in `containMatchingUris` it started working as expected. Before that, sometimes the some images didn't showed up, or in a three columns layout there were just two or one columns with its images and the correspondent width, I mean 1/2 or 1/3 the screen width, respectively.
Is it possible to make some property to enable or disable the cache in the meantime? Thanks!

**Comment by hoscarcito** [comment]
Hi again!
Is there any progress here? I don't know how I can help but I can test, check, etc.
Thanks!

**Comment by brh55** [comment]
@hoscarcito I was trying to figure out the main root issue **(view related v.s. state related)** without having to remove the diff checker. Would you be able to try the `fix-reload` branch and see if this eliminates the bug?

**Comment by hoscarcito** [comment]
It's been a long time, sorry! I used the fix in containMatchingUris and I didn't care about the performance.
I've updated the app, I'll test it!
Thanks for your effort!

**Comment by hoscarcito** [comment]
Hi @brh55 ! I implemented your modifications but nothing changed.
This are the only different lines I found in that branch

          scrollRenderAheadDistance={100}
          removeClippedSubviews={false}

If that's correct, nothing happens, I'm still having the problem.
Sorry! 😞

**Comment by brh55** [comment]
@hoscarcito Appreciate the testing efforts that confirms my assumption with it being an isolated issue with the caching logic. Fortunately, most of these issues are resolved in the upcoming version v0.5.0. Just been having build issues with updating the example to the latest RN that's been forcing me to push back the release date.
